index.js

```diff
@@ -948,75 +973,63 @@ RedisClient.prototype.auth = function () {
+    var field, tmp_args;
+    if (Array.isArray(key)) {
+        return this.send_command("hmset", key, args);
     }
-};
-RedisClient.prototype.HMGET = RedisClient.prototype.hmget;
-
-RedisClient.prototype.hmset = function (args, callback) {
-    var tmp_args, tmp_keys, i, il, key;
-
     if (Array.isArray(args)) {
-        return this.send_command("hmset", args, callback);
-    }
-
-    args = to_array(arguments);
-    if (typeof args[args.length - 1] === "function") {
-        callback = args[args.length - 1];
-        args.length -= 1;
-    } else {
-        callback = null;
+        return this.send_command("hmset", [key].concat(args), callback);
     }
-
-    if (args.length === 2 && (typeof args[0] === "string" || typeof args[0] === "number") && typeof args[1] === "object") {
+    if (typeof args === "object") {
         // User does: client.hmset(key, {key1: val1, key2: val2})
         // assuming key is a string, i.e. email address
 
         // if key is a number, i.e. timestamp, convert to string
-        if (typeof args[0] === "number") {
-            args[0] = args[0].toString();
+        // TODO: This seems random and no other command get's the key converted => either all or none should behave like this
+        if (typeof key !== "string") {
+            key = key.toString();
         }
-
-        tmp_args = [ args[0] ];
-        tmp_keys = Object.keys(args[1]);
-        for (i = 0, il = tmp_keys.length; i < il ; i++) {
-            key = tmp_keys[i];
-            tmp_args.push(key);
-            tmp_args.push(args[1][key]);
+        tmp_args = [key];
+        for (field in args) {
+            tmp_args.push(field, args[field]);
         }
-        args = tmp_args;
+        return this.send_command("hmset", tmp_args, callback);
     }
-
-    return this.send_command("hmset", args, callback);
+    return this.send_command("hmset", to_array(arguments));
 };
-RedisClient.prototype.HMSET = RedisClient.prototype.hmset;
-
-Multi.prototype.hmset = function () {
-    var args = to_array(arguments), tmp_args;
-    if (args.length >= 2 && typeof args[0] === "string" && typeof args[1] === "object") {
-        tmp_args = [ "hmset", args[0] ];
-        Object.keys(args[1]).map(function (key) {
-            tmp_args.push(key);
-            tmp_args.push(args[1][key]);
-        });
-        if (args[2]) {
-            tmp_args.push(args[2]);
+
+Multi.prototype.hmset = Multi.prototype.HMSET = function (key, args, callback) {
+    var tmp_args, field;
+    if (Array.isArray(key)) {
+        if (args) {
+            key.push(args);
+        }
+        tmp_args = ['hmset'].concat(key);
+    } else if (Array.isArray(args)) {
+        if (callback) {
+            args.push(callback);
+        }
+        tmp_args = ['hmset', key].concat(args);
+    } else if (typeof args === "object") {
+        tmp_args = ["hmset", key];
+        if (typeof key !== "string") {
+            key = key.toString();
+        }
+        for (field in args) {
```

Comment:
Using for in without checking for hasOwnProperty is faster than using Object.keys and looping over that. I can't think of a reason why someone would add something to the object prototype if it'll be passed to node_redis.